spec/mem/tree_borrows/state_machine.md

```diff
@@ -44,49 +49,70 @@ impl Permission {
-    fn foreign_read(self) -> Result<Permission> {
-        ret(
+    fn foreign_read(self, protected: Protected, accessed: Accessed) -> Result<Permission> {
+        match self {
+            Permission::Active if protected.active(accessed) =>
+                throw_ub!("Tree Borrows: reading from the foreign of an actively protected pointer with Active permission"),
+            Permission::Reserved { .. } if protected.yes() => ret(Permission::Reserved { conflicted: true }),
+            Permission::Active => ret(Permission::Frozen),
+            _ => ret(self),
+        }
+    }
+
+    fn foreign_write(self, protected: Protected, accessed: Accessed) -> Result<Permission> {
+        if !protected.active(accessed) {
             match self {
-                Permission::Reserved => Permission::Reserved,
-                Permission::Active => Permission::Frozen,
-                Permission::Frozen => Permission::Frozen,
-                Permission::Disabled => Permission::Disabled,
+                Permission::ReservedIM => return ret(self),
+                _ => return ret(Permission::Disabled),
             }
-        )
-    }
+        }
 
-    // FIXME: consider interior mutability
-    fn foreign_write(self) -> Result<Permission> {
-        ret(Permission::Disabled)
+        match self {
+            Permission::Reserved { .. } => throw_ub!("Tree Borrows: writing to the foreign of an actively protected pointer with Reserved permission"),
+            Permission::Active => throw_ub!("Tree Borrows: writing to the foreign of an actively protected pointer with Active permission"),
+            Permission::Frozen => throw_ub!("Tree Borrows: writing to the foreign of an actively protected pointer with Frozen permission"),
+            Permission::Disabled => panic!("Permission::foreign_write: Impossible Combination (Accessed + Protected + Disabled)"),
+            Permission::ReservedIM => panic!("Permission::foreign_write: Impossible Combination (Accessed + Protected + ReservedIM)"),
+        }
     }
 
     fn transition(
         self,
         access_kind: AccessKind,
         node_relation: NodeRelation,
+        protected: Protected,
+        accessed: Accessed,
     ) -> Result<Permission> {
         match (node_relation, access_kind) {
-            (NodeRelation::Foreign, AccessKind::Write) => self.foreign_write(),
-            (NodeRelation::Foreign, AccessKind::Read) => self.foreign_read(),
+            (NodeRelation::Foreign, AccessKind::Write) => self.foreign_write(protected, accessed),
+            (NodeRelation::Foreign, AccessKind::Read) => self.foreign_read(protected, accessed),
+            (NodeRelation::Child, AccessKind::Write) => self.child_write(protected),
             (NodeRelation::Child, AccessKind::Read) => self.child_read(),
-            (NodeRelation::Child, AccessKind::Write) => self.child_write(),
+        }
+    }
+
+    fn default(mutbl: Mutability, pointee: PointeeInfo, protected: Protected) -> Result<Permission> {
```

Comment:
This doesn't need `result`.

Also please move this to the top of the impl block.